### PR TITLE
Lecture 16

### PR DESCRIPTION
**Pull request by mexal17**
Используя java + testNg + Webdriver

1)	 Написать тест который логинится на onliner.by  или любой другой интернет магазин 
2)	Выбирает рандомный продукт 
3)	Кладет в корзину
4)	Заходит в корзину и проверяет наличие выбранного товара

Использовать все виды локаторов : id, name, class, xpath, css

### CODE REVIEW

File: Lecture-16/pom.xml [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+            </plugin>
+        </plugins>
+    </build>
+
+    <dependencies>
+
+        <dependency>
+            <groupId>org.seleniumhq.selenium</groupId>
+            <artifactId>selenium-chrome-driver</artifactId>
+            <version>3.13.0</version>
```
Comment (ivan-pronin) [inline Lecture-16/pom.xml]: 1. Use one version across similar dependencies, e.g. for all `org.seleniumhq.selenium` deps use one version.
2. Use latest available library versions: https://mvnrepository.com/artifact/org.seleniumhq.selenium/selenium-java, so `3.141.59` is ok

File: Lecture-16/src/test/java/OnlinerSiteTest.java [hunk 2]
```diff
@@ -0,0 +1,95 @@
+import org.openqa.selenium.By;
```
Comment (ivan-pronin) [inline Lecture-16/src/test/java/OnlinerSiteTest.java]: package?

File: Lecture-16/src/test/java/OnlinerSiteTest.java [hunk 3]
```diff
@@ -0,0 +1,95 @@
...
+import java.util.Collections;
+import java.util.List;
+import java.util.NoSuchElementException;
+
+public class OnlinerSiteTest {
+    private static final String URL = "https://onliner.by";
+    private static final String REG1 = "//div[@class='auth-bar__item auth-bar__item--text']";
+    private static final String REG2 = "//input[@type='text' and @class='auth-input auth-input_primary auth-input_base auth-form__input auth-form__input_width_full']";
+    private static final String REG3 = "//input[@type='password' and @class='auth-input auth-input_primary auth-input_base auth-form__input auth-form__input_width_full']";
+    private static final String REG4 = "//button[@type='submit' and @class='auth-button auth-button_primary auth-button_middle auth-form__button auth-form__button_width_full']";
```
Comment (ivan-pronin) [inline Lecture-16/src/test/java/OnlinerSiteTest.java]: REG1 ... REG4 - not clear variable names

File: Lecture-16/src/test/java/OnlinerSiteTest.java [hunk 4]
```diff
@@ -0,0 +1,95 @@
...
+import java.time.Duration;
+import java.util.Collections;
+import java.util.List;
+import java.util.NoSuchElementException;
+
+public class OnlinerSiteTest {
+    private static final String URL = "https://onliner.by";
+    private static final String REG1 = "//div[@class='auth-bar__item auth-bar__item--text']";
+    private static final String REG2 = "//input[@type='text' and @class='auth-input auth-input_primary auth-input_base auth-form__input auth-form__input_width_full']";
+    private static final String REG3 = "//input[@type='password' and @class='auth-input auth-input_primary auth-input_base auth-form__input auth-form__input_width_full']";
```
Comment (ivan-pronin) [inline Lecture-16/src/test/java/OnlinerSiteTest.java]: I think `//input[@type='password']` is enough

File: Lecture-16/src/test/java/OnlinerSiteTest.java (hunk 3 above)
Comment (ivan-pronin) [inline Lecture-16/src/test/java/OnlinerSiteTest.java]: I think `//button[@type='submit']` will be enough

File: Lecture-16/src/test/java/OnlinerSiteTest.java [hunk 5]
```diff
@@ -0,0 +1,95 @@
...
+import org.testng.annotations.Test;
+
+import java.time.Duration;
+import java.util.Collections;
+import java.util.List;
+import java.util.NoSuchElementException;
+
+public class OnlinerSiteTest {
+    private static final String URL = "https://onliner.by";
+    private static final String REG1 = "//div[@class='auth-bar__item auth-bar__item--text']";
```
Comment (ivan-pronin) [inline Lecture-16/src/test/java/OnlinerSiteTest.java]: Bad locator - difficult to understand, a lot of classes which may be changed

File: Lecture-16/src/test/java/OnlinerSiteTest.java [hunk 6]
```diff
@@ -0,0 +1,95 @@
...
+    private static final String REG2 = "//input[@type='text' and @class='auth-input auth-input_primary auth-input_base auth-form__input auth-form__input_width_full']";
```
Comment (ivan-pronin) [inline Lecture-16/src/test/java/OnlinerSiteTest.java]: Bad locator - difficult to understand, a lot of classes which may be changed

File: Lecture-16/src/test/java/OnlinerSiteTest.java [hunk 7]
```diff
@@ -0,0 +1,95 @@
...
+    private static final String REG3 = "//input[@type='password' and @class='auth-input auth-input_primary auth-input_base auth-form__input auth-form__input_width_full']";
+    private static final String REG4 = "//button[@type='submit' and @class='auth-button auth-button_primary auth-button_middle auth-form__button auth-form__button_width_full']";
+    private static final String PRODUCTS_XPATH = "//div[@class='schema-product__title']/a";
+    private static final String SECTIONS_CSS_SELECTOR = ".catalog-bar__link.catalog-bar__link_strong";
+    private static final String LOGIN = "Mexal17";
+    private static final String PASS = "123";
+    private static final String CATALOG_XPATH = "//span[text()='Каталог']/..";
+    private static final String SELLERS_CSS_SELECTOR = ".button.button_orange.button_big.offers-description__button";
+    private static final String PUT_INTO_BASKET_CSS_SELECTOR = ".button.button_orange.button_middle.offers-list__button.offers-list__button_basket";
```
Comment (ivan-pronin) [inline Lecture-16/src/test/java/OnlinerSiteTest.java]: Bad locator - difficult to understand, a lot of classes which may be changed

File: Lecture-16/src/test/java/OnlinerSiteTest.java [hunk 8]
```diff
@@ -0,0 +1,95 @@
...
+    private static final String OPEN_BASKET_XPATH = ".button.button_orange.button_middle.offers-list__button.offers-list__button_basket.offers-list__button_checked";
```
Comment (ivan-pronin) [inline Lecture-16/src/test/java/OnlinerSiteTest.java]: Bad locator - difficult to understand, a lot of classes which may be changed

File: Lecture-16/src/test/java/OnlinerSiteTest.java [hunk 9]
```diff
@@ -0,0 +1,95 @@
...
+    private static final String ITEMS_CLASS_NAME = "cart-product";
+    private static final String DELETE_CLASS_NAME = "cart-product__remove";
+    private static final String NO_ITEMS_CSS_SELECTOR = ".cart-message.cart-message_empty";
+    private static final String PROPERTY_KEY = "webdriver.chrome.driver";
```
Comment (ivan-pronin) [inline Lecture-16/src/test/java/OnlinerSiteTest.java]: Not clear var name - what property?

File: Lecture-16/src/test/java/OnlinerSiteTest.java [hunk 10]
```diff
@@ -0,0 +1,95 @@
...
+    private static final String PROPERTY_VALUE = "./src/test/resources/chromedriver.exe";
```
Comment (ivan-pronin) [inline Lecture-16/src/test/java/OnlinerSiteTest.java]: Not clear var name - what value?

File: Lecture-16/src/test/java/OnlinerSiteTest.java [hunk 11]
```diff
@@ -0,0 +1,95 @@
...
+        Duration interval = Duration.ofMillis(10);
+        wait = new FluentWait<WebDriver>(driver).withTimeout(waitTime).pollingEvery(interval).ignoring(NoSuchElementException.class);
+    }
+
+    @Test
+    public void registrationTest() {
+        driver.findElement(By.xpath(REG1)).click();
+        driver.findElement(By.xpath(REG2)).sendKeys(LOGIN);
+        driver.findElement(By.xpath(REG3)).sendKeys(PASS);
+        wait.until(webDriver -> driver.findElement(By.xpath(REG4))).click();
```
Comment (ivan-pronin) [inline Lecture-16/src/test/java/OnlinerSiteTest.java]: Add assertion for successful registration / login

File: Lecture-16/src/test/java/OnlinerSiteTest.java [hunk 12]
```diff
@@ -0,0 +1,95 @@
...
+        wait.until(webDriver -> driver.findElement(By.xpath(CATALOG_XPATH))).click();
+        List<WebElement> sections = wait.until(webDriver -> driver.findElements(By.cssSelector(SECTIONS_CSS_SELECTOR)));
+        Collections.shuffle(sections);
+        sections.get(0).click();
+        System.out.println("sections.size = " + sections.size());
+        List<WebElement> products = wait.until(webDriver -> driver.findElements(By.xpath(PRODUCTS_XPATH)));
+        Collections.shuffle(products);
+        products.get(0).click();
+        System.out.println("products.size() = " + products.size());
+        Assert.assertNotEquals(0, products.size());
```
Comment (ivan-pronin) [inline Lecture-16/src/test/java/OnlinerSiteTest.java]: Useless assertion - if products were empty you got exception at line `products.get(0).click()`

File: Lecture-16/src/test/java/OnlinerSiteTest.java [hunk 13]
```diff
@@ -0,0 +1,95 @@
...
+    }
+
+    @Test(dependsOnMethods = "chooseProductTest")
+    public void putIntoBasketTest() {
+        driver.findElement(By.cssSelector(SELLERS_CSS_SELECTOR)).click();
+        driver.findElement(By.cssSelector(PUT_INTO_BASKET_CSS_SELECTOR)).click();
+        wait.until(webDriver -> driver.findElement(By.cssSelector(OPEN_BASKET_XPATH))).click();
+        List<WebElement> items = wait.until(webDriver -> driver.findElements(By.className(ITEMS_CLASS_NAME)));
```
Comment (ivan-pronin) [inline Lecture-16/src/test/java/OnlinerSiteTest.java]: `ITEMS_CLASS_NAME` - not clear var name

File: Lecture-16/src/test/java/OnlinerSiteTest.java [hunk 14]
```diff
@@ -0,0 +1,95 @@
...
+        System.out.println("items.size() = " + items.size());
+        Assert.assertNotEquals(0, items.size());
```
Comment (ivan-pronin) [inline Lecture-16/src/test/java/OnlinerSiteTest.java]: Use `Assert.assertTrue(!items.isEmpty())` - it's easier to understand or
`Assert.assertFalse(items.isEmpty())`

**Review by ivan-pronin — CHANGES_REQUESTED** [review]

File: Lecture-16/src/test/java/OnlinerSiteTest.java (hunk 3 above)
Comment (mexal17) [reply Lecture-16/src/test/java/OnlinerSiteTest.java]: done

File: Lecture-16/src/test/java/OnlinerSiteTest.java (hunk 4 above)
Comment (mexal17) [reply Lecture-16/src/test/java/OnlinerSiteTest.java]: done

File: Lecture-16/src/test/java/OnlinerSiteTest.java (hunk 3 above)
Comment (mexal17) [reply Lecture-16/src/test/java/OnlinerSiteTest.java]: done

File: Lecture-16/src/test/java/OnlinerSiteTest.java (hunk 5 above)
Comment (mexal17) [reply Lecture-16/src/test/java/OnlinerSiteTest.java]: done

File: Lecture-16/src/test/java/OnlinerSiteTest.java (hunk 6 above)
Comment (mexal17) [reply Lecture-16/src/test/java/OnlinerSiteTest.java]: done

File: Lecture-16/pom.xml (hunk 1 above)
Comment (mexal17) [reply Lecture-16/pom.xml]: done

File: Lecture-16/src/test/java/OnlinerSiteTest.java (hunk 7 above)
Comment (mexal17) [reply Lecture-16/src/test/java/OnlinerSiteTest.java]: done

File: Lecture-16/src/test/java/OnlinerSiteTest.java (hunk 8 above)
Comment (mexal17) [reply Lecture-16/src/test/java/OnlinerSiteTest.java]: done

File: Lecture-16/src/test/java/OnlinerSiteTest.java (hunk 9 above)
Comment (mexal17) [reply Lecture-16/src/test/java/OnlinerSiteTest.java]: done

File: Lecture-16/src/test/java/OnlinerSiteTest.java (hunk 10 above)
Comment (mexal17) [reply Lecture-16/src/test/java/OnlinerSiteTest.java]: done

File: Lecture-16/src/test/java/OnlinerSiteTest.java (hunk 11 above)
Comment (mexal17) [reply Lecture-16/src/test/java/OnlinerSiteTest.java]: done

File: Lecture-16/src/test/java/OnlinerSiteTest.java (hunk 12 above)
Comment (mexal17) [reply Lecture-16/src/test/java/OnlinerSiteTest.java]: done

File: Lecture-16/src/test/java/OnlinerSiteTest.java (hunk 13 above)
Comment (mexal17) [reply Lecture-16/src/test/java/OnlinerSiteTest.java]: done

File: Lecture-16/src/test/java/OnlinerSiteTest.java (hunk 14 above)
Comment (mexal17) [reply Lecture-16/src/test/java/OnlinerSiteTest.java]: done

File: Lecture-16/src/test/java/OnlinerSiteTest.java (hunk 2 above)
Comment (mexal17) [reply Lecture-16/src/test/java/OnlinerSiteTest.java]: done

File: Lecture-16/src/test/java/website/OnlinerSiteTest.java [hunk 15]
```diff
@@ -0,0 +1,106 @@
...
+    private static final String DRIVER_TYPE = "webdriver.chrome.driver";
+    private static final String DRIVER_PATH = "./src/test/resources/chromedriver.exe";
+    private static final String PRODUCTS_FRAME_XPATH = "//main[@class='product-main']";
+
+    private WebDriver driver;
+    private Wait<WebDriver> wait;
+
+    private void clickOnRandomElement(List<WebElement> elements){
+        Collections.shuffle(elements);
+        elements.get(0).click();
```
Comment (ivan-pronin) [inline Lecture-16/src/test/java/website/OnlinerSiteTest.java:47]: `IndexOutOfBoundsException` if `elements.isEmpty()`

File: Lecture-16/src/test/java/website/OnlinerSiteTest.java [hunk 16]
```diff
@@ -0,0 +1,106 @@
...
+    private static final String DELETE_ITEMS_XPATH = "//a[@title='Удалить']";
+    private static final String NO_ITEMS_XPATH = "//a[text()='историю заказов']";
+    private static final String DRIVER_TYPE = "webdriver.chrome.driver";
+    private static final String DRIVER_PATH = "./src/test/resources/chromedriver.exe";
+    private static final String PRODUCTS_FRAME_XPATH = "//main[@class='product-main']";
+
+    private WebDriver driver;
+    private Wait<WebDriver> wait;
+
+    private void clickOnRandomElement(List<WebElement> elements){
```
Comment (ivan-pronin) [inline Lecture-16/src/test/java/website/OnlinerSiteTest.java:45]: This method should be moved to a separate class

File: Lecture-16/src/test/java/website/OnlinerSiteTest.java [hunk 17]
```diff
@@ -0,0 +1,106 @@
...
+        Collections.shuffle(elements);
+        elements.get(0).click();
+    }
+
+    @BeforeClass
+    public void setup() {
+        System.setProperty(DRIVER_TYPE, DRIVER_PATH);
```
Comment (ivan-pronin) [inline Lecture-16/src/test/java/website/OnlinerSiteTest.java:52]: Move all driver related code to `WebDriverManager` class

File: Lecture-16/src/test/java/website/OnlinerSiteTest.java [hunk 18]
```diff
@@ -0,0 +1,106 @@
...
+        driver.manage().window().maximize();
+        driver.get(URL);
+        Duration waitTime = Duration.ofSeconds(10);
+        Duration interval = Duration.ofMillis(100);
+        wait = new FluentWait<WebDriver>(driver).withTimeout(waitTime).pollingEvery(interval).ignoring(NoSuchElementException.class);
+    }
+
+    @Test
+    public void logInTest() {
+        driver.findElement(By.xpath(AUTH_BUTTON)).click();
```
Comment (ivan-pronin) [inline Lecture-16/src/test/java/website/OnlinerSiteTest.java:64]: `NoSuchElementException` if no element is found. Tests should handle such cases proactively

Comment (ivan-pronin) [reply Lecture-16/src/test/java/website/OnlinerSiteTest.java:64]: Extract appropriate method to separate classes - `SearchActions`, `ClickActions`, `SendKeysActions`, `WaitActions` and reuse in all test methods

**Review by ivan-pronin — CHANGES_REQUESTED** [review]

File: Lecture-16/src/test/java/website/OnlinerSiteTest.java (hunk 15 above)
Comment (mexal17) [reply Lecture-16/src/test/java/website/OnlinerSiteTest.java:47]: done

**Pull request merged**